engine/verification/assigner/engine.go

```diff
@@ -63,36 +73,159 @@ func (e *Engine) Done() <-chan struct{} {
+// (through the chunk assigner), and all belong to the same execution result.
+//
+// Deduplication of chunk locators is delegated to the chunks queue.
+func (e *Engine) processChunks(chunkList flow.ChunkList, resultID flow.Identifier) {
+	for _, chunk := range chunkList {
+		log := e.log.With().
+			Hex("result_id", logging.ID(resultID)).
+			Hex("chunk_id", logging.ID(chunk.ID())).
+			Uint64("chunk_index", chunk.Index).Logger()
+
+		locator := &chunks.Locator{
+			ResultID: resultID,
+			Index:    chunk.Index,
+		}
+
+		// pushes chunk locator to the chunks queue
+		ok, err := e.chunksQueue.StoreChunkLocator(locator)
 		if err != nil {
-			engine.LogError(e.log, err)
+			log.Error().Err(err).Msg("could not push chunk locator to chunks queue")
+			continue
+		}
+		if !ok {
+			log.Debug().Msg("could not push duplicate chunk locator to chunks queue")
+			continue
 		}
-	})
+
+		// notifies chunk queue consumer of a new chunk
+		e.newChunkListener.Check()
+		log.Debug().Msg("chunk locator successfully pushed to chunks queue")
+	}
 }
 
-func (e *Engine) ProcessLocal(event interface{}) error {
-	return e.Process(e.me.NodeID(), event)
+// stakedAtBlockID checks whether this instance of verification node has staked at specified block ID.
+// It returns true and nil if verification node is staked at referenced block ID, and returns false and nil otherwise.
+// It returns false and error if it could not extract the stake of node as a verification node at the specified block.
+func (e *Engine) stakedAtBlockID(blockID flow.Identifier) (bool, error) {
+	identity, err := protocol.IdentityAtBlockID(e.state, blockID, e.me.NodeID())
+	if err != nil {
+		return false, fmt.Errorf("could not extract staked identify of node at block %v: %w", blockID, err)
+	}
+
+	if identity.Role != flow.RoleVerification {
+		return false, fmt.Errorf("node is staked for an invalid role. expected: %s, got: %s", flow.RoleVerification, identity.Role)
+	}
+
+	staked := identity.Stake > 0
+	return staked, nil
 }
 
-func (e *Engine) Process(originID flow.Identifier, event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(originID, event)
-	})
+// ProcessFinalizedBlock indexes the execution receipts included in the block, and handles their chunk assignments.
+// Once it is done handling all the receipts in the block, it notifies the block consumer.
+func (e *Engine) ProcessFinalizedBlock(block *flow.Block) {
+	blockID := block.ID()
+	log := e.log.With().
+		Hex("block_id", logging.ID(blockID)).
+		Int("receipt_num", len(block.Payload.Receipts)).Logger()
+
+	log.Debug().Msg("new finalized block arrived")
+
+	err := e.indexer.IndexReceipts(blockID)
+	if err != nil {
+		// TODO: consider aborting the process
+		log.Error().Err(err).Msg("could not index receipts for block")
+	}
+
+	for _, receipt := range block.Payload.Receipts {
+		e.handleExecutionReceipt(receipt, blockID)
```

Comment:
we could let it return the error, and reuse the same logger above to log with the context.

and `handlExecutionReceipt` don't need to log, but only return error. 

`handleExecutionReceipt` should only return fatal error right? In that case, we could handle it with `e.log.Fatal`

Reply:
To address this comment, I tried refactoring `handleExecutionReceipt` and `processChunk` to be invoked as follows:
`chunkList, staked, err := e.handleExecutionReceipt(receipt)`
and
`ok, err := e.processChunks(chunk, resultID)`

However, after refactoring, I found the code is degraded on readability and encapsulation, due to nested loggings: 
As the result of the refactoring `ProcessFinalizedBlock` would look like this:
```
	blockID := block.ID()
	
	// FIRST LEVEL LOGGER. 
	log := e.log.With().
		Hex("container_block_id", logging.ID(blockID)).
		Uint64("container_block_height", block.Header.Height).
		Int("receipt_num", len(block.Payload.Receipts)).Logger()

	log.Debug().Msg("new finalized block arrived")

	err := e.indexer.IndexReceipts(blockID)
	if err != nil {
		log.Error().Err(err).Msg("could not index receipts for block")
	}

	for _, receipt := range block.Payload.Receipts {
		resultID := receipt.ExecutionResult.ID()
		
		// SECOND LEVEL LOGGER (extended from first level one)
		log := log.With().
			Hex("receipt_id", logging.Entity(receipt)).
			Hex("result_id", logging.ID(receipt.ExecutionResult.ID())).
			Hex("reference_block_id", logging.ID(receipt.ExecutionResult.BlockID)).
			Logger()

		chunkList, staked, err := e.handleExecutionReceipt(receipt)
		if err != nil {
			// we should log smt here
			continue
		}
		if !staked {
			// we should log smt here
			continue
		}

		// we should log that we have handled the receipt and the number of assigned chunks. 

		// pushes assigned chunks to chunks queue.
		for _, chunk := range chunkList {
                         // THIRD LEVEL LOGGER (extended from the second one). 
			log := log.With().
				Hex("chunk_id", logging.ID(chunk.ID())).
				Uint64("chunk_index", chunk.Index).Logger()

			ok, err := e.processChunks(chunk, resultID)
			if err != nil {
			     // we should log smt here
			     continue
			}

			if !ok {
                            // we should log smt here
                            continue
			}
			
			// we should log that we processed this chunk
		}
```

Here are what I observed on this:
1- The `ProcessFinalizedBlock` is getting unnecessarily longer and less readable compared to what it is right now. 
2- We should be very careful about using `continue` which was not the case previously -> harder to maintain. 
3- We have these 3-level depth nested logging that should be extra cautious about their scope, of course, we can keep defining local variables (e.g., `receiptLogger` and `chunkLogger`) to add more distinction, nevertheless, we degrade the readability and maintainability of our code. Additionally, the initial purpose of having the logger as a member field of `engine` and hence shared among all its methods is faded in this approach. 

**So to address this comment based on the observations:**
1- Refactored `processChunks` to operate on a single chunk instead of a chunk list (so that we can use `return` and instead of `continue`). 
2- Kept handling logs locally to the methods for more encapsulation and readability. 
3- Escalated the log levels to `fatal` in cases we are not expecting an error (e.g., storing receipt in database). 

Of course any suggestion on this is welcomed and I can try those in possibly further PRs.